EventProc/src/OverlayProducer.cxx

```diff
@@ -0,0 +1,332 @@
+	  nBunchesToSample_    = parameters.getParameter< int >("nBunchesToSample");
+	  bunchSpacing_        = parameters.getParameter< double >("bunchSpacing");
+	  
+      verbosity_           = parameters.getParameter< int >("verbosity");
+
+	  /// Print the parameters actually set. Helpful in case of typos.
+      if (verbosity_) {
+		 ldmx_log(info) <<  "Got parameters \n \t overlayFileName = " << overlayFileName_ 
+				  << "\n\t sim pass name = " << simPassName_
+				  << "\n\t overlay pass name = " << overlayPassName_
+				  << "\n\t overlayHitCollections = " ;
+		for ( const std::string &coll : collections_ )
+		   ldmx_log(info) <<  coll << "; " ;
+
+		ldmx_log(info) 
+				  << "\n\t numberOverlaidInteractions = " << poissonMu_
+				  << "\n\t doPoisson = " << doPoisson_
+				  << "\n\t timeSpread = " << timeSigma_
+				  << "\n\t timeMean = " << timeMean_
+				  << "\n\t verbosity = " << verbosity_ ;
+		
+      }
+	  
+	  return;
+    }
+  
+  void OverlayProducer::produce(Event& event) {
+	//event is the incoming, simulated event/"hard" process
+	//overlayEvent_ is the overlay producer's own event. 
+
+	if (verbosity_ > 1) {
+	   ldmx_log(info) <<  "produce() starts on simulation event " <<  event.getEventHeader().getEventNumber() ;
+	}
+
+	/// set up random seeds
+	if (rndm_.get() == nullptr) {
+	  //not been seeded yet, get it from RNSS
+	  const auto& rnss = getCondition<RandomNumberSeedService>(RandomNumberSeedService::CONDITIONS_OBJECT_NAME);
+	  rndm_ = std::make_unique<TRandom2>(rnss.getSeed("OverlayProducer::rndm"));
+	}
+	if (rndmTime_.get() == nullptr) {
+	  //not been seeded yet, get it from RNSS
+	  const auto& rnss = getCondition<RandomNumberSeedService>(RandomNumberSeedService::CONDITIONS_OBJECT_NAME);
+	  rndmTime_ = std::make_unique<TRandom2>(rnss.getSeed("OverlayProducer::rndmTime"));
+	}
+
+	
+	// sample a poisson distribution, or use a deterministic number of overlay events
+	int nEvsOverlay = doPoisson_ ? (int) rndm_->Poisson( poissonMu_ ) : (int)poissonMu_ ;
+	// the poisson samples the total number of events, which is nOverlay + 1 (since it includes the sim event)
+	nEvsOverlay -=1; // now subtract the sim event from the poisson mu
+
+	
+	if (verbosity_ > 2) {
+	   ldmx_log(debug) <<  "will overlay " <<  nEvsOverlay << " events on the simulated one" ;
+	}
+	
+ 	
+	//get event wherever nextEvent()  left us 
+	if (! &overlayEvent_ ) {
+	  ldmx_log(error) << "No overlay event!" ;
+	  return;
+	}
+	
+	if (verbosity_ > 2) {
+	   ldmx_log(debug) <<  "starting from overlay event " <<  overlayEvent_.getEventHeader().getEventNumber() ;
+	}
+
+      
+	// using nextEvent to loop, we need to loop over overlay events and in an inner loop, loop over collections, and store them.
+	// after all pileup events have been added, the vector of collections is iterated over and added to the event bus.
+	
+	std::vector< std::vector<SimCalorimeterHit> >  v_outHits;
+	std::map< std::string, int > collectionMap;
```

Comment:
Is there a reason that you don't use
`std::map< std::string , std::vector<SimCalorimeterHit> > collectionMap`
?
This would seem to simplify your code somewhat... I will include other comments that offer an alternative using this setup.

Reply:
No reason. I thought about it too but 1. had a vague recollection of more complicated object maps not being allowed (probably an older standard?) and 2. was mostly happy when I got things to work and considered this to be  cosmetics. I'll implement it when I get back to this code (once I have some help on the SimTrackerHits implementation).